paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/AbstractDataTableSource.java

```diff
@@ -214,7 +214,7 @@ private DataStream<RowData> configureSource(
             FlinkSourceBuilder sourceBuilder, StreamExecutionEnvironment env) {
         Options options = Options.fromMap(this.table.options());
         Integer parallelism = options.get(FlinkConnectorOptions.SCAN_PARALLELISM);
-        if (options.get(FlinkConnectorOptions.INFER_SCAN_PARALLELISM)) {
+        if (parallelism == null && options.get(FlinkConnectorOptions.INFER_SCAN_PARALLELISM)) {
```

Comment:
It seems that we did not perform boundary check(e.g. `>0`) on `parallelism`. If `SCAN_PARALLELISM` is set to `-1`, `OperatorValidationUtils.validateParallelism` will not reject it because it is equal to `ExecutionConfig.PARALLELISM_DEFAULT`, which may cause some confusion. But it can always be argued that this is the behavior supported by `flink` (-1 is considered the default parallelism). Just a comment, maybe not do this check here is exactly the right one. 🤔

Reply:
Yes, let this behavior explained by Flink.